### ISSUE TITLE
macOS - VoiceOver / Chrome announcing visually hidden text out of order

### ISSUE THREAD

**Issue by joe-watkins** [h5bp/html5-boilerplate#1985]
**I'm submitting a ...**

- [x] bug report
- [ ] feature request
- [ ] other (Please do not submit support requests here (below))

## Steps to reproduce
1. Visit https://codepen.io/joe-watkins/pen/OjpqxL with Safari or Chrome and VoiceOver activated.
2. Tab to each link and listen to how they are announced.

## Expected Results
The visually hidden text using HTML5 boilerplate's `.visuallyhidden {}` will be announced in the order that it is within the markup.

## Actual Results
The order of the text is not announced does not match the markup order. Visually hidden text is announced before visible text copy.

## Potential fix
After some testing, I've found that using `position: relative;` and `display: inline-block;` appears to solve this. e.g.

```
.visuallyhidden {
	border: 0;
	clip: rect(0 0 0 0);
	clip-path: inset(50%);
	width: 1px;
	height: 1px;
	margin: -1px;
	overflow: hidden;
	padding: 0;
	position: relative; // different - for reading order in macOS VO
	display: inline-block; // new - for reading order in macOS VO
}
```
## SR Tested (works in)
- Win Edge,IE11,FF,Chrome JAWS 18/17
- Win IE11,FF,Chrome NVDA
- macOS Seirra Safari,Chrome,FF VoiceOver

## Browsers tested (works - hides text in)
- Win Edge, IE11, IE10, IE9, IE8, FF 53, Chrome 60
- macOS Seirra Safari 10.2, Chrome 60, FF 54
 
## Video screencast of experience
[![Everything Is AWESOME](https://img.youtube.com/vi/qRk5gkKg4Ys/0.jpg)](https://www.youtube.com/watch?v=qRk5gkKg4Ys "Video screencast of visually hidden text and VoiceOver reading out of order")

I'd love to get some thoughts on this and I'd be happy to create a PR if everyone is into it :)

**Comment by scottaohara** [comment]
Doing some [additional testing on this](https://codepen.io/scottohara/pen/LjyYbm) and I've identified that it's the negative margin that is specifically to blame for the incorrect reading order.

If you either 1. delete the negative margin or 2. set the margin to zero, the correct reading order is reinstated.  

Awesome find @joe-watkins

**Comment by roblarsen** [comment]
🥇 bug report @joe-watkins 

Thanks for bringing this to our attention. 

Changes to this class will require a ton of testing (which you've already started 👍) to make sure there are no regressions in other situations. I'd also _really_ like to understand the underlying issue.

**Comment by joe-watkins** [comment]
@scottaohara Aww interesting -- I'm sure that negative margin is there for a reason... can anyone remember?

@roblarsen tnx! :) Glad to help. Reading order is being affected by visual presentation some how. Very odd. It sure would be less heavy handed to look at the `margin` declaration per @scottaohara 
 's finds rather than changing `position` and adding `display`.

But, I'm wondering if it may be advantageous to have less absolutely positioned elements in the DOM? That a perf thing?

It feels silly to attack an Apple bug from this angle but we know how bug submissions and updates go with that camp.

**Comment by roblarsen** [comment]
@joe-watkins it's an [OLD commit](https://github.com/h5bp/html5-boilerplate/commit/07768edc28ac7843f230ab617c58ff1fd9dd49c1) relating to issue #194, where there is a LOT of discussion of the original code. I'll dive in a bit later to check it out myself, but there's got to be some reasoning in there.

**Comment by scottaohara** [comment]
I assumed the negative margin was for legacy browsers (since the linked commit/issue reference ie7), as I have yet to find a situation where it's been useful/needed at all in ie9+ projects.

doing a quick check in ie7/8/9, i'm still trying to find exactly what that margin did... possibly a fall back to negate a stray pixel if the visuallyhidden content had a background color?

**Comment by joe-watkins** [comment]
@scottaohara Yes, after poking around a bit it looks like it is to combat single pixel appearing.

I wonder if this dark art would be considered:

``` CSS
.visuallyhidden {
  ...
  margin: -1px/9; // ie8 and below
  ...
}
```

**Comment by scottaohara** [comment]
while i am a fan of the occasional practicing of dark arts, @joe-watkins, the README states it supports ie9+

I haven't seen the pixel in ie9 plus, have you with your testing?

**Comment by joe-watkins** [comment]
@scottaohara Nor have I seen any issues in IE9... even by giving it a huge width/height and background color. `clip` is doing the magic with the visibility.

Removal of the negative margin sounds like the fix.

Thoughts?

**Comment by roblarsen** [comment]
We don't need to worry about old IE anymore, so if we can clean this up with the removal of the negative margin, then that's great. I'd love to see a PR for this. One last bug fix before I ship 6.0.

**Comment by roblarsen** [comment]
Closed via #1986

**Comment by joe-watkins** [comment]
Noticed issues persists when a `line-height` is defined. Continued testing needed.

https://github.com/h5bp/html5-boilerplate/pull/1986#issuecomment-322253065

**Comment by scottaohara** [comment]
[my comment outlining additional options to rectify the bug](https://github.com/h5bp/html5-boilerplate/pull/1986#issuecomment-322271340)

**Comment by roblarsen** [comment]
BUMMER.

The good news is, removing old IE code is still a valid thing to do these days.

**Comment by joe-watkins** [comment]
Hey all, so after some fiddling around with the help of @scottaohara in https://github.com/h5bp/html5-boilerplate/pull/1986 we have landed on a solution a bit more like what I originally was thinking with a twist..bringing back the negative margin which helps with visually oddities with focus ring, removing the absolute positioning, and introducing `display: inline-block;` and things appear to be working well.

I've introduced this snippet into existing projects and things hold up well in all browser/at combinations I can test (mentioned in original post). Reading order in Safari/Chrome/VoiceOver is correct.

It would be great if some others could run this through the ringer:

``` CSS
.visuallyhidden {
    border: 0;
    clip: rect(0 0 0 0);
    clip-path: inset(50%);
    display: inline-block;
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    width: 1px;
    white-space: nowrap; /* 1 */
}
```

**Comment by joe-watkins** [comment]
As pointed out by @tomasz1986 in https://github.com/h5bp/html5-boilerplate/pull/1986#issuecomment-322661868 the absolute positioning was structural for the `clip` for old browsers. Recommending removal of that. 

``` CSS
.visuallyhidden {
    border: 0;
    clip-path: inset(50%);
    display: inline-block;
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    width: 1px;
    white-space: nowrap;
}
```

https://codepen.io/joe-watkins/pen/vJWMLp

With a green light I could submit a PR for this :)

**Comment by roblarsen** [comment]
What testing has been done? Is there any missing coverage?

**Comment by joe-watkins** [comment]
Hi @roblarsen below you will find the browsers/AT I've tested the latest CSS with and all is well.

## SR/AT Tested (works as expected in)
Win IE11,IE10,FF,Chrome with JAWS 18/17
Win IE11 w/Dragon Naturally Speaking
Win IE11,IE10,FF,Chrome NVDA
macOS Seirra Safari,Chrome,FF VoiceOver
iOS 10 VoiceOver / Safari
Android 7 TalkBack / Chrome

## Browsers tested (works - visually hides text in)
Win Edge, IE11, IE10, IE9, IE8, IE6, FF 53, Chrome 60
macOS Seirra Safari 10.2, Chrome 60, FF 54
iOS 10 VO / Safari

**Comment by roblarsen** [comment]
Thanks @joe-watkins, that's perfect

**Comment by roblarsen** [comment]
closed via #1989

**Comment by svinkle** [comment]
Bad news: without `position: absolute` the `.visuallyhidden` element adds space to the container.

Demo: https://codepen.io/svinkle/pen/LjQJbZ

/cc @joe-watkins @scottaohara @roblarsen

**Comment by scottaohara** [comment]
(╯°□°）╯︵ ┻━┻

**Comment by svinkle** [comment]
I mean, you _could_ apply a work-around to the extra space by specifying a container `width` and `height` but that's not realistic. 😕

**Comment by joe-watkins** [comment]
@svinkle Ahh nuts! Looks like a code/whitespace thing: https://codepen.io/joe-watkins/pen/yovRzJ (hmmmff!)

**Comment by roblarsen** [comment]
This issue is like a see-saw.

**Comment by svinkle** [comment]
@joe-watkins Interesting…

@roblarsen I might suggest a roll-back on the latest merge until this is sorted. Having extra, visual space is worse than the read order bug, imo.

**Comment by scottaohara** [comment]
so as @joe-watkins said, it's definitely related to the `display: inline-block;` causing [extra white space](https://css-tricks.com/fighting-the-space-between-inline-block-elements/).

initial testing of setting margin to `margin: -1px -1px -1px -.25em;` seems to be doing the trick right now: https://codepen.io/scottohara/pen/GvQYOe?editors=1100

but would love additional hammering please

**Comment by joe-watkins** [comment]
This is in no way a fix.. but does appear to address this case. `float` Needs further testing.

https://codepen.io/joe-watkins/pen/YxeJJB?editors=1100

Perhaps we can't fix a Webkit/Blink bug from this angle and need to revert?

**Comment by svinkle** [comment]
`float` is causing more/different spacing issues on the site I'm working on…

I vote revert.

**Comment by scottaohara** [comment]
yeh.  revert is probably the way to go, but humor me and give this a shot, @svinkle 

```
.visuallyhidden-with-out-height {
    border: 0 !important;
    clip: rect(0 0 0 0);
    clip-path: inset(50%);
    overflow: hidden;
    width: 1px;
    position: absolute;
    white-space: nowrap;
}
```

**Comment by svinkle** [comment]
@scottaohara No dice; reads out of order. 😕

**Comment by scottaohara** [comment]
time to re-file a bug with webkit

**Comment by joe-watkins** [comment]
👍 on revert.

**Comment by joe-watkins** [comment]
I've submitted two bugs. I know @svinkle has submitted one with Apple in the past.

**Chrome:**
https://bugs.chromium.org/p/chromium/issues/detail?id=757177

**Apple:**
Bug ID: 33979302

@roblarsen We may need to revert back to before #1986

**Comment by scottaohara** [comment]
Do we want to revert to before or after the removal of the negative margin?  
`margin: 0;` clearly wasn't the fix we needed, but it did help for some use cases & didn't introduce any other bugs.

Thoughts?

**Comment by svinkle** [comment]
I'd say revert to abc64efcb71847049620bf8ae78f3cd5ce26bf5e. The class definition was quite stable at this point in time.

I can't say for certain, but pretty sure `margin: -1px` exists to offset the `width: 1px` and `height: 1px` properties.

**Comment by roblarsen** [comment]
I'm going to revert to the stable version later today

**Comment by scottaohara** [comment]
@svinkle right, it offsets it, but from what we've identified, that seems to be part of the issue, offsetting it outside of the focus ring area.

being that it's a position absolute element, the margin doesn't affect the DOM, the clip/clip-path are what actually hide the element. this just seems to be arbitrarily moving it, which may have had a purpose in older browsers, but other versions of the visuallyhidden class don't use this declaration at all anymore.  

Apologies if this comes across as me being obstinate, that is not my intent and I agree reverting to a stable version is the way to go. I'm just trying to figure out what the benefit of this particular declaration is.  I'll just keep testing :)

**Comment by svinkle** [comment]
@scottaohara Sounds good to me! If we're certain it serves no purpose and doesn't negatively effect the class, by all means. 👍

**Comment by scottaohara** [comment]
@svinkle @joe-watkins 

I continue to battle against this...
I've included multiple tests and am trying out various versions of the hiding content classes.

Presently I'm using the 
.at-only class that I often use myself and have modified a bit (current version in this codepen reads content in order, in the testing i'm currently doing)
.visuallyhidden from html5bp (without margin)
.snooka from https://snook.ca/archives/html_and_css/hiding-content-for-accessibility 

please feel free to join me if you want to waste your time as well :)

**Comment by KayLeung** [comment]
As I remember, the current code was fixed the `width: 0; height:0` problem for VoiceOver [ref: https://snook.ca/archives/html_and_css/hiding-content-for-accessibility]. I tested the latest MacOS and has no this issue. Is it the right time back to a simple workaround with ZERO value?

Could anyone test the older MacOS?

**Comment by bradkemper** [comment]
Has anyone tried `margin: 0 -1px -1px 0`? I think that would prevent offsetting content by the 1px width and height, without influencing the width and height of the container, nor the reading order. Assuming left to right reading.

**Comment by afercia** [comment]
For reference, similar issue in WordPress:
https://core.trac.wordpress.org/ticket/42006

**Comment by fixframe** [comment]
good job

**Comment by fixframe** [comment]
good job

**Comment by Malvoz** [comment]
For some _inspiration_ (read: copy/paste), this is how [Google AMP](https://github.com/ampproject/amphtml/blob/master/css/amp.css) does it:

```
/**
 * Makes elements visually invisible but still accessible to screen-readers.
 *
 * This Css has been carefully tested to ensure screen-readers can read and
 * activate (in case of links and buttons) the elements with this class. Please
 * use caution when changing anything, even seemingly safe ones. For example
 * changing width from 1 to 0 would prevent TalkBack from activating (clicking)
 * buttons despite TalkBack reading them just fine. This is because
 * element needs to have a defined size and be on viewport otherwise TalkBack
 * does not allow activation of buttons.
 */
.i-amphtml-screen-reader {
  position: fixed !important;
  /* keep it on viewport */
  top: 0px !important;
  left: 0px !important;
  /* give it non-zero size, VoiceOver on Safari requires at least 2 pixels
     before allowing buttons to be activated. */
  width: 4px !important;
  height: 4px !important;
  /* visually hide it with overflow and opacity */
  opacity: 0 !important;
  overflow: hidden !important;
  /* remove any margin or padding */
  border: none !important;
  margin: 0 !important;
  padding: 0 !important;
  /* ensure no other style sets display to none */
  display: block !important;
  visibility: visible !important;
}
```

**Comment by svinkle** [comment]
Thanks for pointing this out, @Malvoz. Unfortunately the bug persists when testing with @joe-watkins's original pen here: https://codepen.io/joe-watkins/pen/OjpqxL 😕

These days I'm using `aria-label` more than `.visually-hidden` elements if I can help it. Support is 💯and ensures a properly formatted label in the accessibility tree.

**Comment by joe-watkins** [comment]
@Malvoz Interesting,.. I hadn't experimented much defining a size of 4px. Such a weird bug. I wish Safari wasn't such a walled garden.

@svinkle Dagnabbit! I can still get `aria-label` to fail in current Safari / VO on an anchor still when using only the up/down arrow keys for navigation. https://s.codepen.io/joe-watkins/debug/QMpPQQ

**Comment by scottaohara** [comment]
@joe-watkins yeh, Safari + VO don't respect ARIA as much when using up/down arrows to navigate content.

aria-labels and aria-hidden are not respected at all, just to name a couple.  Up/down arrows allow me to exit the majority of modal dialog scrips I've come across :)

But back to the mention of width, if you recall when we were hammering on this, I found that if the width is not restricted, this issue can be partially mitigated.

**Comment by afercia** [comment]
> @joe-watkins yeh, Safari + VO don't respect ARIA as much **when using up/down arrows** to navigate content.

> aria-labels and aria-hidden are not respected at all, just to name a couple. Up/down arrows allow me to exit the majority of modal dialog scrips I've come across :)

Maybe because with Safari and VoiceOver you have to use `Control-Option-Right Arrow` and `Control-Option-Left Arrow` to navigate content? ;) `aria-label` and `aria-hidden` work perfectly for me and I haven't heard of any big issues with them.

**Comment by scottaohara** [comment]
That's one way to navigate content.  but navigating line by line can be done with up/down arrows.  so *have to* is not exactly an accurate statement there.

**Comment by afercia** [comment]
As well as "aria-labels and aria-hidden are not respected at all" :)

**Comment by joe-watkins** [comment]
@afercia Yes using VO keys things behave as expected, but if you don't use the VO keys and only the arrow keys `aria-label` fails.. so the alternative method for navigating is an option and there isn't a way to enforce one method of navigation over another. Quick nav behaves well which is good :)

**Comment by afercia** [comment]
@joe-watkins fair enough :) my point is that in VoiceOver `only the arrow keys` works only in Quick Nav (which I guess it's rarely used), otherwise you have to use the VO modifier. This is a different model compared to the one used by screen readers on Windows and users are aware of it.
So there's no great point in saying "aria-labels and aria-hidden are not respected at all" when actually using the wrong keystrokes. That's just a wrong information.

https://help.apple.com/voiceover/info/guide/10.12/#/vo27974

**Comment by scottaohara** [comment]
It's a different way for a user to navigate a document, yes, but it's not "wrong information".  if a user uses VO key + arrow keys to navigate content and they have one experience, and then they use up/down arrow keys to navigate content and they experience different content announcements that may not be clear to them as to why.

the fact of the matter is that an aria-label will not be read out when using up/down arrows, as Joe has pointed out with his demo.  aria-hidden content *will* be announced if navigating in this manner.  regardless of if the controls are "right wrong" it's what happens, so stating it's "wrong" information is in itself "wrong".

**Comment by Malvoz** [comment]
@aferica, @scottaohara

Concerning navigational methods - as [**spatial navigation**](https://github.com/WICG/spatial-navigation) is on the rise, using only arrow keys to navigate web content will most likely be a much more common scenario in the near future.

**Comment by constancecchen** [comment]
One thing to consider about `aria-label` as well is that [it's currently ignored by Google and Microsoft's auto translators](http://www.heydonworks.com/article/aria-label-is-a-xenophobe), making for worse internationalization.

(This has been a super fascinating discussion to read by the way, thanks all for the very thorough testing and documentation!)

**Comment by roblarsen** [comment]
> (This has been a super fascinating discussion to read by the way, thanks all for the very thorough testing and documentation!)
 
Seconded! Community discussion, input and testing are how projects like this thrive and this issue (even though we don't have a solution) is a perfect example.

**Comment by scottaohara** [comment]
It's a really good point @constancecchen 

I've been thinking about this since reading your comment yesterday, and I have a hack for this hack and i'm just going to say hack again because it's a hack.  But, this is the only thing I've come up with that negates the reordering while also taking your comments about internationalization into account:

```html
<style>
   [hidden][aria-hidden="true"] { display: inline; }
</style>
<a href="#!">
  <span hidden aria-hidden="true" >Read More</span>
  <span class="visuallyhidden">Read more about such and such</span> 
</a>
```

So breaking down that less than ideal markup...
1. instead of using an `aria-label`we take all the content that would have been its value, and stick it into a `visuallyhidden` `span`.  Now it can be translated and since there's no breaks in the text, it will be read as intended.
2. Set the visible text to `aria-hidden='true'` so that it's ignored by VO when the link is tabbed to, or focused by VO+left/right arrow navigation.  Now there won't be duplication in announcements.
3. Set the visible "read more" content to be hidden by the HTML hidden attribute. Re-show it with CSS undoing the attribute's default `display none`.  This is done specifically for no CSS situations (e.g. a browser's reader mode) to mitigate the possibility of "Read more Read more about ...."

With all that said, this seems like it's probably more trouble than its worth?

Also of note, in testing this with Safari and Firefox's reader mode, I had been using a `sr-only` class on my content and Firefox's reader mode style sheet specifically targets that class (and `visually-hidden` but not `visuallyhidden`) and sets those classes to `display: none;`.  Safari seems to target `sr-only` content and simply doesn't carry it over into their reader mode version of the document at all.

Switching my class name from `sr-only` to `at-only` allowed the visually hidden content to be carried over to each browser's reader mode... so per what I outlined above "Read more about such and such" was available on screen, but the `[hidden]` "read more" text was not present.

Seems like there may be more wrinkles to using this class/technique than one might think...

**Comment by joe-watkins** [comment]
@scottaohara @roblarsen @constancecchen Ha! This issue rules. I've not considered translation services and aria-labels nor reader mode w/visually hidden text. Love it!

@scottaohara I think we could make an argument for filing bugs with browsers on *NOT* using `display: none` for those classes that they think are used for hiding content from screen readers. SR w/Reader mode may be edge case - but accessibility is all about reach!

Firefox styles for reader mode (aboutReader.css):
``` CSS
.moz-reader-content .visually-hidden, .moz-reader-content .visuallyhidden, .moz-reader-content .hidden, .moz-reader-content .invisible, .moz-reader-content .sr-only {
  display: none;
}
```

[Bug created w/FF about maybe respecting visually hidden approaches in Reader](https://bugzilla.mozilla.org/show_bug.cgi?id=1464858)

One could probably author a suite of fragile hacks for reader modes that leverage descending selectors.

*.visuallyhidden VO reading order fix??*
Well here goes another stab at the reading order for Safari/VO. This so far, seems to be working. No extra/funky space in focus outline and reads in order. Changed height definition to `em` and tweaked margins to `.25em -25em`. I'll continue to pound at this.. can others take it for a spin?

[CodePen using em-based height and em-based margins](https://s.codepen.io/joe-watkins/debug/wXwmYw)

``` CSS
.visuallyhidden {
	border: 0;
	clip: rect(0 0 0 0);
	height: 1em; /* new - was 1px */
	margin: .25em -.25em; /* new - was -1px */
	overflow: hidden;
	padding: 0;
	position: absolute;
	width: 1px;
	white-space: nowrap; /* 1 */
}
```

**Comment by scottaohara** [comment]
oh yeah @joe-watkins, i completely agree.  filing bugs is by far the way to go.  

and per your new CSS setup, I added loads of content to the visually hidden span, and it announced in the appropriate order...so that's a plus and far better than the hack I posted :)

**Comment by svinkle** [comment]
@joe-watkins Can we change to `rem` units instead of `em` in order to size from the root and not the parent container? I haven't tested but seems like a safer route to me.

**Comment by joe-watkins** [comment]
@svinkle `rem` could work too. I was leaning `em` for wider browser support but I just tried IE6 and everything looks great. Updated to `rem`. Need to test some rwd / typography size changes on root to make sure Safari / VO behaves.

``` CSS
.visuallyhidden {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1rem; /* new - was 1px */
    margin: .25rem -.25rem; /* new - was -1px */
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px;
    white-space: nowrap; /* 1 */
}
```

**Comment by afercia** [comment]
Having faced this issue in various projects, I'd tend to think the root cause is that Safari + VoiceOver tend to read content "by line", or based on what they consider lines of text. For some reason, CSS positioning comes into play when they decide what a "line of text" is.

I've personally used the same CSS fix in some cases, with good results. Giving the visually hidden element the same approximate height of the line of text it's placed onto certainly helps. In a way, the previous text and the visually hidden text will be considered "on the same line". 

However, this fix won't cover all the cases. For example, buttons or other elements containing text might have a height that depends on other CSS properties (explicitly set height, padding, whatever). n my experience, in these cases (see for example the screenshot below) using the 1em/rem fix wouldn't help so much:

<img width="648" alt="6b619f2e-31d2-11e7-9d7a-b2e8040c5d59" src="https://user-images.githubusercontent.com/1682452/40630951-a44fd6ca-62d7-11e8-9ea7-5c74e6cc0754.png">

It would be very interesting if anyone could test the scenario above, when they have a chance 🙂

**Comment by joe-watkins** [comment]
@afercia In my experience/testing, the reading order isn't affected when `<button>` is used in Safari/VO.. now links that look like buttons are a different story :) 

So I've created another demo with an assortment of link/buttons that rely on different heights and such. I noticed the little visual offset creep back in w/Safari's focus ring and tweaked the margin to 0 and that helped. I'm not able to get it to break with these examples.. is there another scenario you are referencing? Can you create a demo of it? tnx!

*updated w/zero margin*/
``` CSS
.visuallyhidden {
	border: 0;
	clip: rect(0 0 0 0);
	height: 1rem; /* new - was 1px */
	margin: 0; /* new - was -1px */
	overflow: hidden;
	padding: 0;
	position: absolute;
	width: 1px;
	white-space: nowrap; /* 1 */
}
```

**Comment by joe-watkins** [comment]
In Safari, with the current CSS there is funkiness with the focus ring. It falls out of bounds a bit.

![blip-current](https://user-images.githubusercontent.com/3695795/40633104-3eca50ee-62a2-11e8-9717-217f2424cc64.png)

With the new/fix using `height: 1rem;` & `margin: .25rem -.25rem` reading order is fixed and the funkiness moves to the bottom of an inline link.

![blip-new-25](https://user-images.githubusercontent.com/3695795/40633158-b2f6508a-62a2-11e8-8774-b7cb9dfde7d6.png)

By using `height: 1rem;` & `margin: 0` the focus funkiness is fixed and reading order is fixed in Safari/VO.

![blip-new-0](https://user-images.githubusercontent.com/3695795/40633184-e16feb24-62a2-11e8-97d7-9132b9eb38c5.png)

But here's the deal, it can be different depending on stuff like resets, Normalize, Bootstrap. The focus funkiness position can be different if using those and I've noticed the `height: 1rem;` & `margin: .25rem -.25rem` work best there.

So with `margin` being a bit squishy what should we set that to? `0` and let authors tweak it? Or base it off popular things like Normalize and set it to `.25rem -.25rem`?

**Comment by afercia** [comment]
@joe-watkins well anything that alters the height of a line of text (meaning the line-height) will probably make it fail. For example, see https://codepen.io/afercia/full/YvzPyB/

<img width="773" alt="screen shot 2018-05-29 at 10 25 59" src="https://user-images.githubusercontent.com/1682452/40646961-f95aea58-632a-11e8-97be-309b461be87f.png">

**Comment by afercia** [comment]
This seems related: https://bugs.webkit.org/show_bug.cgi?id=173914
(although it mentions containers with aria-label or aria-labelledby)

**Comment by joe-watkins** [comment]
@afercia Thanks for putting that demo together.. yes that line height weakness is one that sure makes it fail. I think I've found an approach for that which appears to be holding up... of course I think I've said that 10 times in this thread :) ha! I love a good challenge though.

`height: auto;` to the rescue.

Initial testing of a [clone of your CodePen](https://s.codepen.io/joe-watkins/debug/WyQgwL) seems to be doing well with. Take a peek.

**Comment by joe-watkins** [comment]
@afercia Ha nice finds and makes sense -- especially 8.3 as I was fiddling around with transforms at one point and adding backgrounds to the element and was surprised to see it span the height of the box. Thanks for surfacing those! 

Yes, I've been thinking about flex and grid and wonder if we should be attacking from the parent since we can't drive up the tree with CSS really. One could almost write a visually hidden framework - ha!

I've battled the Flex parent issue with `position: relative;` on the parent, added `.has-flex-parent` to the `.visuallyhidden` element and used `right: 0;` on the hidden element and VO appears to reads in order. Not ideal, but in this case works. Definitely a case-by-case thing with newer CSS technologies.

[View the CodePen demo battling a Flex parent](https://codepen.io/joe-watkins/pen/VdeeJx/)

In related good news, I've learned that in flex/grid scenarios, visually hidden elements that are grandchildren of Flex parents and Grid parents announce in order w/VO+Safari when using the my latest chunk of `.visuallyhidden` CSS.

[Flexbox demo with grandchildren](https://codepen.io/joe-watkins/pen/ERPPMW)
[CSS Grid demo with grandchildren](https://codepen.io/joe-watkins/pen/vrLLwr)

**Comment by nickcolley** [comment]
Hey everyone,

I've been doing some investigation into this since we ran into the same issue on the GOV.UK Design System.

I have written up the results of an audit I've done here: https://gist.github.com/nickcolley/19b80ed24d0364cfd3afd3b1b49c4014

> ## Conclusion
> We should update the govuk-visually-hidden class to remove the negative margin which will fix the ordering issue in VoiceOver for OSX.
> 
> We should recommend using aria-label or aria-labelledby where it makes sense to.
> 
> Visually hidden CSS can result in text not being read out when touching on iOS, this is regrettable but does not cause a hard barrier as users can explore the surrounding content to understand context.

The funkiness with the outline ring which seems to a blocker here is only seen in OSX Safari when you are using the user-agent outline styles.

Since we set an outline style this is not an issue for us. I think we could make the assumption that most people will style their outlines instead of using the default.

So we could update the class here to have the improved announcement for screen readers with the rendering issue and add some guidance that explains the need to style your own outlines (which is good practice regardless).

Hope this is helpful, thanks for your help!

Nick

**Comment by afercia** [comment]
We're facing an issue with flex items, where:
- a button has `display: inline-flex`
- the screen reader text within the button is placed differently across browsers
- specifically, Chrome places the screen reader text _before_ the content and this doesn't seem correct to me: in absence of left/right/top/bottom values, an absolutely positioned element should stay in its original position

For more details: https://github.com/WordPress/gutenberg/pull/13221#issuecomment-453016911. In the past, there have been several inconsistencies across browsers regarding absolutely positioned flex items. I'd greatly appreciate feedback from all the experts here 🙂 can this issue be considered a Chrome bug?
